contracts/token/ERC20/extensions/ERC4626.sol

```diff
@@ -26,15 +26,24 @@ import "../../../utils/math/Math.sol";
 abstract contract ERC4626 is ERC20, IERC4626 {
     using Math for uint256;
 
-    IERC20Metadata private immutable _asset;
+    IERC20 private immutable _asset;
 
     /**
      * @dev Set the underlying asset contract. This must be an ERC20-compatible contract (ERC20 or ERC777).
      */
-    constructor(IERC20Metadata asset_) {
+    constructor(IERC20 asset_) {
         _asset = asset_;
     }
 
+    /** @dev See {IERC4626-asset}. */
+    function decimals() public view virtual override(IERC20Metadata, ERC20) returns (uint8) {
+        try IERC20Metadata(address(_asset)).decimals() returns (uint8 value) {
```

Comment:
I have some serious doubts about making this dynamic. I feel it introduces attack vectors. What do you think about reading the asset decimals in the constructor and storing it as an immutable?

Reply:
That would be better for the vanilla version, but worst for the the upgradable version :/

Reply:
Attack vector seems limited to the underlying asset, which can already attack the vault anyway. I don't see any context where the vault would not trust the underlying token contract.